transformers/modeling_bert.py

```diff
@@ -172,6 +174,67 @@ def forward(self, input_ids, token_type_ids=None, position_ids=None):
         return embeddings
 
 
+class BertGeneralAttention(nn.Module):
+    def __init__(self, config):
+        super(BertGeneralAttention, self).__init__()
+        if config.hidden_size % config.num_attention_heads != 0:
+            raise ValueError(
+                "The hidden size (%d) is not a multiple of the number of attention "
+                "heads (%d)" % (config.hidden_size, config.num_attention_heads))
+        self.output_attentions = config.output_attentions
+
+        self.num_attention_heads = config.num_attention_heads
+        self.attention_head_size = int(config.hidden_size / config.num_attention_heads)
+        self.all_head_size = self.num_attention_heads * self.attention_head_size
+
+        self.query = nn.Linear(config.hidden_size, self.all_head_size)
+        self.key = nn.Linear(config.hidden_size, self.all_head_size)
+        self.value = nn.Linear(config.hidden_size, self.all_head_size)
+
+        self.dropout = nn.Dropout(config.attention_probs_dropout_prob)
+
+    def transpose_for_scores(self, x):
+        new_x_shape = x.size()[:-1] + (self.num_attention_heads, self.attention_head_size)
+        x = x.view(*new_x_shape)
+        return x.permute(0, 2, 1, 3)
+
+    def forward(self, query, key, value, attention_mask=None, head_mask=None):
```

Comment:
Do you think you can just modify the `BertSelfAttention` class with an additional `is_decoder` flag in the configuration and an `encoder_hidden_state=None` input in the forward pass (added at the end to avoid breaking change for people using positional indexing to access the keyword input?

Reply:
I agree that in the short term it would be easier to add a config flag and pass an `encoder_hidden_state` variable to `BertSelfAttention`. Let me explain my rationale for duplicating. The advantage of duplication is that we can progressively deprecate the classes in the internals that are too specific: 

1. Immediately use `BertGeneralAttention` in `BertAttention`, and `BertSelfAttention` would remain only for retro-compatibility with other projects until completely deprecated.
2. We could bypass `BertAttention` in the same way by using the more general `BertDecoderAttention` in `BertEncoderLayer` while keeping it in the code for compatibility. In could rename it `BertGeneralAttention`.

As a result the distinctions between encoder and decoder are contained in the way  `BertEncoderLayer` and `BertDecoderLayer` are built and the way the attention layer is called, while maintaining compatibility. We can further add deprecation warnings when `BertSelfAttention` and `BertAttention` are called from outside the module, and isolate them in the code with appropriate comments.

Now I understand if you would rather stick with the "simple" solution. It's your call.

Reply:
Yes, try to stick with the "simple" solution if you can.
The goal of this project is not to "fix" or "rewrite" the current Bert model but to build on top of it with minimal changes.

Reply:
Will do!